File: android/core/calc/src/main/kotlin/com/stocktracker/core/calc/RowSorting.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.PortfolioRow
import com.stocktracker.core.model.SortField
import com.stocktracker.core.model.SortOrder

/**
 * Orders rows for display.
 *
 * Monetary fields are converted to [displayCurrency] before they are compared: sorting a
 * mixed CZK/USD portfolio on raw values compares numbers in different units, so 100 USD would
 * rank below 200 CZK.
 *
 * Ties break on ticker so the ordering is total — otherwise two rows with the same value
 * could swap places between recompositions.
 *
 * Named `sortedForDisplay` rather than `sortedBy` because the latter is a stdlib extension on
 * Iterable and would be shadowed at every call site.
 */
fun List<PortfolioRow>.sortedForDisplay(
    order: SortOrder,
    displayCurrency: String,
    rates: Map<String, Double>,
): List<PortfolioRow> {
    fun money(amount: Double, from: String) = convert(amount, from, displayCurrency, rates)
    val comparator: Comparator<PortfolioRow> = when (order.field) {
        SortField.NAME -> compareBy { it.name.lowercase() }
        SortField.TYPE -> compareBy { it.type.ordinal }
        SortField.VALUE -> compareBy { money(it.currentValue, it.currency) }
        // Already in the display currency, and it is the figure the row shows.
        SortField.TODAY -> compareBy { it.dailyChangeDisplay }
        SortField.TOTAL_RETURN -> compareBy { money(it.totalReturn, it.currency) }
    }
    val directed = if (order.ascending) comparator else comparator.reversed()
    return sortedWith(directed.thenBy { it.ticker })
}

File: src/utils/rowSorting.ts
import type { PortfolioRow } from '../types'
import type { Convert } from './rowDerivation'

/** The sort fields Android offers; names match `SortField` in core/model/SortOrder.kt. */
export type SortField = 'VALUE' | 'TODAY' | 'TOTAL_RETURN'

/**
 * Orders rows the way Android's `sortedForDisplay` does: money converted to the display
 * currency before comparing, today's change read from the anchored figure (already in the
 * display currency), ties broken on ticker so the order is total.
 */
export function sortRowsForDisplay(
  rows: PortfolioRow[], field: SortField, ascending: boolean, displayCurrency: string, convert: Convert,
): PortfolioRow[] {
  const key = (r: PortfolioRow): number => {
    switch (field) {
      case 'VALUE': return convert(r.currentValue, r.currency, displayCurrency)
      case 'TODAY': return r.dailyChangeDisplay
      case 'TOTAL_RETURN': return convert(r.totalReturn, r.currency, displayCurrency)
    }
  }
  const sign = ascending ? 1 : -1
  return [...rows].sort((a, b) => sign * (key(a) - key(b)) || a.ticker.localeCompare(b.ticker))
}

/**
 * Portfolio today's change: the sum of the anchored per-row figures, which are already in the
 * display currency. Mirrors `portfolioDailyChange` in core/calc/PortfolioSummary.kt.
 */
export function portfolioDailyChange(rows: PortfolioRow[], totalValueDisplay: number): { change: number; percent: number } {
  const change = rows.reduce((s, r) => s + r.dailyChangeDisplay, 0)
  const prev = totalValueDisplay - change
  return { change, percent: prev > 0 ? (change / prev) * 100 : 0 }
}
